File: Python/ps_baekjoon/14503/main.py
import sys

readline, write = (sys.stdin.readline, sys.stdout.write)
cnt = 0
robot = [0] * 2
dr = {
    0: [0, +1, 0, -1],
    1: [-1, 0, +1, 0],
    2: [0, -1, 0, +1],
    3: [+1, 0, -1, 0],
}
dc = {
    0: [-1, 0, +1, 0],
    1: [0, -1, 0, +1],
    2: [+1, 0, -1, 0],
    3: [0, +1, 0, -1],
}
n, m = map(int, readline().strip().split())
robot[0], robot[1], direction = map(int, readline().strip().split())
terrain = [list(map(int, readline().strip().split()))[:m] for i in range(n)]
vst = [[False] * m for i in range(n)]
vst[robot[0]][robot[1]] = True
cnt += 1
while True:
    r, c = robot
    flag = False
    for i in range(4):
        nr = r + dr[direction][i]
        nc = c + dc[direction][i]
        if not (0 <= nr < n and 0 <= nc < m) or terrain[nr][nc] == 1 or vst[nr][nc]:
            continue
        robot[0], robot[1] = nr, nc
        vst[nr][nc] = True
        cnt += 1
        turn = i + 1
        for _ in range(turn):
            direction -= 1
            direction = direction if direction >= 0 else 3
        flag = True
        break
    if not flag:
        nr = r + dr[direction][1]
        nc = c + dc[direction][1]
        if not (0 <= nr < n and 0 <= nc < m) or terrain[nr][nc] == 1:
            break
        robot[0], robot[1] = nr, nc
write(f"{cnt}")

File: C++/ps_baekjoon/14503/main.cpp
#include <bits/stdc++.h>
using namespace std;
using pos = pair<int, int>;

constexpr int mxNM = 50;
int n, m, cnt, direction, terrain[mxNM][mxNM];
bool vst[mxNM][mxNM];
pos robot;
map<int, array<int, 4>>
    dr{
        {0, {0, +1, 0, -1}},
        {1, {-1, 0, +1, 0}},
        {2, {0, -1, 0, +1}},
        {3, {+1, 0, -1, 0}},
    },
    dc{
        {0, {-1, 0, +1, 0}},
        {1, {0, -1, 0, +1}},
        {2, {+1, 0, -1, 0}},
        {3, {0, +1, 0, -1}},
    };

int main()
{
    cin.tie(0)->sync_with_stdio(0);
    cin >> n >> m >> robot.first >> robot.second >> direction;
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < m; ++j)
            cin >> terrain[i][j];
    vst[robot.first][robot.second] = true;
    ++cnt;
    while (true)
    {
        auto [r, c] = robot;
        bool flag = false;
        for (int i = 0; i < 4; ++i)
        {
            int nr = r + dr[direction][i];
            int nc = c + dc[direction][i];
            if (!(0 <= nr && nr < n && 0 <= nc && nc < m) || terrain[nr][nc] == 1 || vst[nr][nc])
                continue;
            robot = {nr, nc};
            vst[nr][nc] = true;
            ++cnt;
            int turn = i + 1;
            while (turn--)
                if (--direction == -1)
                    direction = 3;
            flag = true;
            break;
        }
        if (!flag)
        {
            int nr = r + dr[direction][1];
            int nc = c + dc[direction][1];
            if (!(0 <= nr && nr < n && 0 <= nc && nc < m) || terrain[nr][nc] == 1)
                break;
            robot = {nr, nc};
        }
    }
    cout << cnt;
    return 0;
}